src/main/scala/com/evvo/agent/func.scala

```diff
@@ -23,24 +23,83 @@ abstract class CreatorFunction[Sol](name: String) extends NamedFunction(name) {
+abstract class ModifierFunction[Sol](name: String,
+                                     val numInputs: Int = 32,
+                                     val shouldRunWithPartialInput: Boolean = true)
   extends NamedFunction(name) {
   /**
     * @return The function to call to produce new solutions
     */
-  def mutate(sols: IndexedSeq[Scored[Sol]]): TraversableOnce[Sol]
+  def modify(sols: IndexedSeq[Scored[Sol]]): TraversableOnce[Sol]
+}
+
+/**
+  * A function that uses a one-to-one mapping to derive a new set of solutions from some
+  * input set of solutions.
+  */
+abstract class MutatorFunction[Sol](name: String,
+                                    numInputs: Int = 32)
+  extends ModifierFunction[Sol](name, numInputs, shouldRunWithPartialInput = true) {
+  // Pass shouldRunWithPartialInput as true always - we know there is no requirement for more
+  // than 1 solution at a time, so safe to run on any number of solutions.
+
+  override final def modify(sols: IndexedSeq[Scored[Sol]]): TraversableOnce[Sol] = {
+    // We can define modify in terms of mutation, and ask implementors to just provide a one-to-one
+    // mutator function to map over the solutions.
+    sols.map(_.solution).map(mutate)
+  }
+
+  /**
+    * The method to apply to create new solutions.
+    *
+    * @param sol The solution to base the new solution on.
+    * @return The new solution, to be added to the population.
+    */
+  protected def mutate(sol: Sol): Sol
 }
 
+/**
+  * A function that uses a one-to-one mapping to derive a new set of solutions from some
+  * input set of solutions.
+  */
+abstract class CrossoverFunction[Sol](name: String,
+                                      numInputs: Int = 32)
+  extends ModifierFunction[Sol](name, numInputs, shouldRunWithPartialInput = true) {
+  // Pass shouldRunWithPartialInput as true always - we know there is no requirement for more
+  // than 2 solutions at a time, so safe to run on, say, 16 or 23 solutions. We will have to be
+  // careful about odd numbers of solutions.
+
+  override final def modify(sols: IndexedSeq[Scored[Sol]]): TraversableOnce[Sol] = {
+    // We can define modify in terms of crossover, and ask implementors to just provide a two-to-one
+    // crossover function to map over the solutions.
+    val groups = sols.map(_.solution).grouped(2)
+
+    // Use `collect` and partial functions to ensure that if there is an odd number of solutions,
+    // we never hit a runtime error, like we would with indexing into each group.
+    groups.collect {
+      case IndexedSeq(sol1, sol2) => crossover(sol1, sol2)
+    }
+  }
+
+  /**
+    * Combines two solutions to produce a new one.
+    *
+    * @param sol1 One solution. Order is random, and has no meaning.
+    * @param sol2 Another solution. Order is random, and has no meaning.
+    * @return The new solution, to be added to the population.
+    */
+  protected def crossover(sol1: Sol, sol2: Sol): Sol
```

Comment:
I think this should be consistent with `ModifierFunction` and take `Scored[Sol]`s, and just change modify to set `val groups = sols.grouped(2)`